core/src/main/java/com/linecorp/armeria/common/stream/PathStreamMessage.java

```diff
@@ -0,0 +1,368 @@
+            this.withPooledObjects = withPooledObjects;
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0L) {
+                cancel();
+                downstream.onError(
+                        new IllegalArgumentException("Rule §3.9 violated: non-positive subscription requests " +
+                                                     "are forbidden."));
+            } else {
+                request0(n);
+            }
+        }
+
+        private void request0(long n) {
+            final long requested = this.requested;
+            if (requested == Long.MAX_VALUE) {
+                return;
+            }
+            if (n == Long.MAX_VALUE) {
+                this.requested = Long.MAX_VALUE;
+            } else {
+                this.requested = LongMath.saturatedAdd(requested, n);
+            }
+
+            if (requested > 0) {
+                // PathSubscription is reading a file.
+                // New requests will be handled by 'completed(Integer, ByteBuf)'.
+                return;
+            }
+
+            read();
+        }
+
+        private void read() {
+            if (!reading && !closed && requested > 0) {
+                requested--;
+                reading = true;
+                final ByteBuf buffer = alloc.buffer(bufferSize);
+                fileChannel.read(buffer.nioBuffer(0, bufferSize), position, buffer, this);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            if (executor.inEventLoop()) {
+                cancel0();
+            } else {
+                executor.execute(this::cancel0);
+            }
+        }
+
+        private void cancel0() {
+            if (closed) {
+                return;
+            }
+            closed = true;
+
+            if (!reading) {
+                maybeCloseFileChannel();
+            }
+
+            final CancelledSubscriptionException cause = CancelledSubscriptionException.get();
+            if (notifyCancellation) {
+                downstream.onError(cause);
+            }
+            completionFuture.completeExceptionally(cause);
+            downstream = NoopSubscriber.get();
+        }
+
+        @Override
+        public void completed(Integer result, ByteBuf byteBuf) {
+            executor.execute(() -> {
+                if (closed) {
+                    byteBuf.release();
+                    maybeCloseFileChannel();
+                } else {
+                    if (result > -1) {
+                        position += result;
+                        byteBuf.writerIndex(result);
+                        final HttpData data;
+                        if (withPooledObjects) {
+                            data = HttpData.wrap(byteBuf);
+                        } else {
+                            data = HttpData.wrap(ByteBufUtil.getBytes(byteBuf));
+                            byteBuf.release();
+                        }
```

Comment:
Could we do this without updating `reader/writerIndex`?

Reply:
IIUC, we need to call `byteBuf.writerIndex(result)` because the indexes of `byteBuf` are not changed by `ByteBuffer`.
https://github.com/netty/netty/blob/e5951d46fc89db507ba7d2968d2ede26378f0b04/buffer/src/main/java/io/netty/buffer/ByteBuf.java#L2292-L2294